subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipUpToDateStep.java

```diff
@@ -62,6 +63,8 @@ public UpToDateResult execute(C context) {
                 if (LOGGER.isInfoEnabled()) {
                     LOGGER.info("Skipping {} as it is up-to-date.", context.getWork().getDisplayName());
                 }
+                @SuppressWarnings("OptionalGetWithoutIsPresent")
+                AfterPreviousExecutionState afterPreviousExecutionState = context.getAfterPreviousExecutionState().get();
```

Comment:
I really don't like these calls. I think keeping the previous state in `ExecutionStateChanges` was a better solutin.

Reply:
What are "these calls"? I prefer calling `get` here, knowing that I have a previous execution instead of having `getPreviousExecution` throw an `UnsupportedOperationException` in `RebuildExecutionStateChanges`.

In both cases the type system doesn't help me if I am allowed to obtain the previous execution. When querying from the `context`, I at least am reminded that I might be doing something wrong.

Reply:
I also really didn't like `UOE`s. I'm just saying this doesn't seem like an improvement to me. It also doesn't make things worse I guess. Have we got a plan to make it better yet?

Reply:
The only way to make it better would be to encode in the type system that when you have no changes, then you had a history. I don't know how we would do this. Do you have an idea?